cli/js/compiler.ts

```diff
@@ -81,259 +81,282 @@ interface CompileResult {
+    bundle,
+    target,
+    writeFile
+  }));
+  let diagnostics: readonly ts.Diagnostic[] | undefined;
+
+  // if there is a configuration supplied, we need to parse that
+  if (config && config.length && configPath) {
+    const configResult = host.configure(configPath, config);
+    diagnostics = processConfigureResponse(configResult, configPath);
+  }
+
+  // This will recursively analyse all the code for other imports,
+  // requesting those from the privileged side, populating the in memory
+  // cache which will be used by the host, before resolving.
+  const resolvedRootModules = await processImports(
+    rootNames.map(rootName => [rootName, rootName]),
+    undefined,
+    bundle || host.getCompilationSettings().checkJs
+  );
+
+  let emitSkipped = true;
+  // if there was a configuration and no diagnostics with it, we will continue
+  // to generate the program and possibly emit it.
+  if (!diagnostics || (diagnostics && diagnostics.length === 0)) {
+    const options = host.getCompilationSettings();
+    const program = ts.createProgram({
+      rootNames,
+      options,
+      host,
+      oldProgram: TS_SNAPSHOT_PROGRAM
+    });
+
+    diagnostics = ts
+      .getPreEmitDiagnostics(program)
+      .filter(({ code }) => !ignoredDiagnostics.includes(code));
+
+    // We will only proceed with the emit if there are no diagnostics.
+    if (diagnostics && diagnostics.length === 0) {
+      if (bundle) {
+        // we only support a single root module when bundling
+        assert(resolvedRootModules.length === 1);
+        // warning so it goes to stderr instead of stdout
+        console.warn(`Bundling "${resolvedRootModules[0]}"`);
+        setRootExports(program, resolvedRootModules[0]);
       }
+      const emitResult = program.emit();
+      emitSkipped = emitResult.emitSkipped;
+      // emitResult.diagnostics is `readonly` in TS3.5+ and can't be assigned
+      // without casting.
+      diagnostics = emitResult.diagnostics;
+    }
+  }
+
+  const result: CompileResult = {
+    emitSkipped,
+    diagnostics: diagnostics.length
+      ? fromTypeScriptDiagnostic(diagnostics)
+      : undefined
+  };
 
-      // This will recursively analyse all the code for other imports,
-      // requesting those from the privileged side, populating the in memory
-      // cache which will be used by the host, before resolving.
-      const resolvedRootModules = await processImports(
-        rootNames.map(rootName => [rootName, rootName]),
+  util.log("<<< compile end", {
+    rootNames,
+    type: CompilerRequestType[request.type]
+  });
+
+  return result;
+}
+
+async function requestRuntimeCompile(
+  request: CompilerRequestRuntimeCompile
+): Promise<RuntimeCompileResult | RuntimeBundleResult> {
+  // `RuntimeCompile` are requests from a runtime user, both compiles and
+  // bundles.  The process is similar to a request from the privileged
+  // side, but also returns the output to the on message.
```

Comment:
This comment would make a good jsdoc.

I feel like the function would be better called `runtimeCompile` or `compileRuntime`

Reply:
`compileRuntime` would mean compiling the runtime, where `runtimeCompile` sounds right to me.  I agree with dropping the `request`, as these are imperatives, not sort of "pretty please do something" requests.

Reply:
Also +1 on JSDoc! :-)